### ISSUE TITLE
Changing display mode is now delayed by loading.

### ISSUE THREAD

**Issue by GoogleFrog** [beyond-all-reason/spring#192]
The lobby library in chobby has a callin OnBattleAboutToStart which it sends just before a game is started via any method (with Spring.Reload etc). The main purpose of the callin is to let the settings widget write springsettings.cfg so that the settings are sure to be used in the game. It also changes the display mode to the type selected for ingame display mode. Players can have a windowed lobby and fullscreen game to support hardcore IRC mode.
![image](https://user-images.githubusercontent.com/1318249/150705894-b71149a0-727f-42b1-a027-6b45597a0ba4.png)

On 1544 the display mode switch would fullscreen spring as the game started loading. On 798, with the old lobby code, spring would stay as a window until loading is complete, then fullscreen once in the game. This is a problem for many reasons:
- Loading in a tiny window makes the loadscreen tiny and looks highly unpolished.
- It makes the minimap load at a low resolution and look terrible when blown up to fullscreen.
- It breaks UI loading. Widgets think they don't have enough screen space. Widgets think that the centre of the screen is offset.

To workaround this I added a delay of 0.3 seconds between triggering OnBattleAboutToStart and actually starting the battle. I loaded in a small window just now (after loading properly about 20 times over the past few days) so just increased the delay to 0.45 seconds. This solution is nowhere near as bad as loading in a small window, but it still has problems:
- Blowing up the lobby and waiting half a second looks unpolished. The 1544 behaviour just looked right because the load transition and display mode transition happened at the same time.
- A delay for an event that is not allowed to fail (eg making a connection to a game) is a ticking timebomb in the codebase, just waiting to appear as a weird race condition or state invalidation bug.

It's not an urgent thing to solve, especially if the current tradeoff is for reduced bugs with changing display mode, but it is something to keep in mind when revisiting display modes.

**Comment by lhog** [comment]
Window change happens one draw frame after something was requested for change. The rationale is following: every change of springsetting param related to window geometry is triggering the immediate feedback otherwise, causing flickering and unwanted resource reallocation on ViewResize(). As there is no real way to tell when a batch of configuration changes is over, the window is just marked changed and the actual action is deferred one frame after. Usually it's not an issue, except probably your case.

I found no other way to stay compatible to springsettings and not cause flickering, but I'm happy to hear other ideas how to satisfy both conditions.

**Comment by GoogleFrog** [comment]
Is it guaranteed to be one frame? If so then I can improve my hax.

**Comment by lhog** [comment]
Yes it should be guaranteed. Some stages of the game might still not increment drawframe counter,  but it shouldn't be the case.

Also ping me tomorrow I'll make a build that has no change delay. Let's see if flickering is better than delay.

**Comment by WatchTheFort** [comment]
@GoogleFrog I needed to use a two drawframe delay when resizing a bordered window to take into account the window borders. One frame to change to a bordered window, and one frame for Spring to update the border sizes returned in `Spring.GetWindowGeometry`

**Comment by lhog** [comment]
https://github.com/beyond-all-reason/spring/commit/2b45e548a97a5c3f6727f8d4ea61f0717882f7ae and https://github.com/beyond-all-reason/spring/commit/99dd38323d68f51f3e71b40f1e6d30eb1594e243 when polished will hopefully provide some improvements.
